### PR TITLE
feat: add double-comparison lint with tests

### LINKED ISSUE #9: Add `double_comparison`
# What it does

Checks for double comparisons that could be simplified to a single expression.

# Why is this bad?

Readability.

# Example

```rs
if x == y || x < y {}
```
Use instead:
```rs
if x <= y {}
```
https://rust-lang.github.io/rust-clippy/master/index.html#/double_comparisons

### PR DESCRIPTION
**Pull request by renzobanegass**
Closes #9 

**Changes Introduced**
Added double_comparison Lint:

Implemented in plugin.rs to identify cases where double comparisons can be simplified.
The lint detects patterns where a variable or expression is compared twice with == and < operators, suggesting a simplification to a single <= operator.
Implemented Automatic Fix:

Added functionality in fix.rs to automatically correct identified double comparisons.
The fix simplifies the code by replacing x == y || x < y with x <= y, improving code readability and reducing redundancy.
Test Coverage:

Added comprehensive tests to ensure the lint correctly identifies and simplifies double comparisons.
Test cases include:
- Simple double comparisons.
- Complex expressions with arithmetic operations.
- Cases with no double comparisons (to verify the lint does not trigger unnecessarily).
- Edge cases with different variables to ensure the lint is not overly aggressive.

Please tell me if there's any more tests that I should add, and if I implemented the lint properly or there's something missing.

### CODE REVIEW

File: crates/cairo-lint-core/src/fix.rs [hunk 1]
```diff
@@ -62,10 +62,57 @@ impl Fixer {
...
+                                | (BinaryOperator::GE(_), BinaryOperator::GT(_))
+                                | (BinaryOperator::LE(_), BinaryOperator::LT(_))
+                        ) {
+                            let simplified_op = match (lhs_inner.op(db), rhs_inner.op(db)) {
+                                (BinaryOperator::EqEq(_), BinaryOperator::LT(_))
+                                | (BinaryOperator::LT(_), BinaryOperator::EqEq(_))
+                                | (BinaryOperator::LE(_), BinaryOperator::LT(_))
+                                | (BinaryOperator::GE(_), BinaryOperator::GT(_)) => "<=",
+                                _ => return node.get_text(db).to_string(),
+                            };
```
Comment (0xLucqs) [inline crates/cairo-lint-core/src/fix.rs:98]: This looks redundant you can remove the first `matches!` and directly do a match. Also i think you're missing a few patterns. For reference [here](https://github.com/rust-lang/rust-clippy/blob/5da97d006e33a4458f467e41ca43085f5e5b20ba/clippy_lints/src/operators/double_comparison.rs#L38-L50) is the `clippy` implementation

File: crates/cairo-lint-core/src/fix.rs [hunk 2]
```diff
@@ -62,10 +62,57 @@ impl Fixer {
...
+                }
+            }
+        }
+        node.get_text(db).to_string()
+    }
+
+    pub fn extract_variable(binary_expr: &ExprBinary, db: &dyn SyntaxGroup) -> Option<String> {
+        let lhs = binary_expr.lhs(db);
+        Some(lhs.as_syntax_node().get_text(db).to_string())
+    }
```
Comment (0xLucqs) [inline crates/cairo-lint-core/src/fix.rs:117]: why do you return an option here ?

File: crates/cairo-lint-core/src/plugin.rs [hunk 3]
```diff
@@ -20,17 +22,55 @@ pub struct CairoLint;
...
+            let lhs = binary_op.lhs(db);
+            let rhs = binary_op.rhs(db);
+
+            if let (Some(lhs_op), Some(rhs_op)) =
+                (self.extract_binary_operator(&lhs, db), self.extract_binary_operator(&rhs, db))
+            {
+                if (matches!(lhs_op, BinaryOperator::EqEq(_)) && matches!(rhs_op, BinaryOperator::LT(_)))
+                    || (matches!(lhs_op, BinaryOperator::LT(_)) && matches!(rhs_op, BinaryOperator::EqEq(_)))
+                    || (matches!(lhs_op, BinaryOperator::GE(_)) && matches!(rhs_op, BinaryOperator::GT(_)))
+                    || (matches!(lhs_op, BinaryOperator::LE(_)) && matches!(rhs_op, BinaryOperator::LT(_)))
```
Comment (0xLucqs) [inline crates/cairo-lint-core/src/plugin.rs:57]: check [here](https://github.com/rust-lang/rust-clippy/blob/5da97d006e33a4458f467e41ca43085f5e5b20ba/clippy_lints/src/operators/double_comparison.rs#L38-L50) to get all the patterns

File: crates/cairo-lint-core/src/plugin.rs [hunk 4]
```diff
@@ -20,17 +22,55 @@ pub struct CairoLint;
...
 
+impl CairoLint {
+    const DESTRUCT_MATCH: &'static str =
+        "you seem to be trying to use `match` for destructuring a single pattern. Consider using `if let`";
+    const MATCH_FOR_EQUALITY: &'static str =
+        "you seem to be trying to use `match` for an equality check. Consider using `if`";
+    const DOUBLE_COMPARISON: &'static str =
+        "redundant double comparison found. Consider simplifying to a single comparison.";
+
+    pub fn check_double_comparison(&self, db: &dyn SyntaxGroup, expr: &Expr, diagnostics: &mut Vec<PluginDiagnostic>) {
```
Comment (0xLucqs) [inline crates/cairo-lint-core/src/plugin.rs:46]: this should move into its own file

File: crates/cairo-lint-core/tests/test_files/double_comparison/double_comparison [hunk 5]
```diff
@@ -0,0 +1,317 @@
+//! > complex double comparison
```
Comment (0xLucqs) [inline crates/cairo-lint-core/tests/test_files/double_comparison/double_comparison:1]: don't work with felts here use `u32` (`felt252` doesn't derive `PartialOrd`)

File: crates/cairo-lint-core/tests/test_files/double_comparison/double_comparison [hunk 6]
```diff
@@ -0,0 +1,317 @@
...
+ --> lib.cairo:5:24
+    if (x + z) == y || (x + z) < y {
+                       ^*********^
+
+//! > fixed
+fn main() -> bool {
+    let x = 5;
+    let y = 10;
+    let z = 3;
+    if (x + z) == y || (x + z) < y {
```
Comment (0xLucqs) [inline crates/cairo-lint-core/tests/test_files/double_comparison/double_comparison:31]: this should be transformed

**Review by 0xLucqs — CHANGES_REQUESTED** [review]

File: crates/cairo-lint-core/src/fix.rs [hunk 7]
```diff
@@ -232,4 +235,55 @@ impl Fixer {
...
+                let rhs_var = Self::extract_variable(rhs_inner, db);
+
+                if lhs_var == rhs_var {
+                    let op = binary_op.op(db);
+
+                    let simplified_op = match (op, lhs_inner.op(db), rhs_inner.op(db)) {
+                        (BinaryOperator::Or(_), BinaryOperator::EqEq(_), BinaryOperator::LT(_))
+                        | (BinaryOperator::Or(_), BinaryOperator::LT(_), BinaryOperator::EqEq(_))
+                        | (BinaryOperator::Or(_), BinaryOperator::LE(_), BinaryOperator::LT(_))
+                        | (BinaryOperator::Or(_), BinaryOperator::GE(_), BinaryOperator::GT(_)) => "<=",
```
Comment (0xLucqs) [inline crates/cairo-lint-core/src/fix.rs:257]: `x >= y || x > y` isn't equal to `x <= y`
```suggestion
                        | (BinaryOperator::Or(_), BinaryOperator::LE(_), BinaryOperator::LT(_)) => "<=",
```

File: crates/cairo-lint-core/src/fix.rs [hunk 8]
```diff
@@ -232,4 +235,55 @@ impl Fixer {
...
+
+                        (BinaryOperator::Or(_), BinaryOperator::EqEq(_), BinaryOperator::GT(_))
+                        | (BinaryOperator::Or(_), BinaryOperator::GT(_), BinaryOperator::EqEq(_))
+                        | (BinaryOperator::Or(_), BinaryOperator::GE(_), BinaryOperator::LE(_)) => ">=",
```
Comment (0xLucqs) [inline crates/cairo-lint-core/src/fix.rs:261]:
```suggestion
                        | (BinaryOperator::Or(_), BinaryOperator::GT(_), BinaryOperator::GE(_))
                        | (BinaryOperator::Or(_), BinaryOperator::GE(_), BinaryOperator::GT(_)) => ">=",
```

File: crates/cairo-lint-core/src/fix.rs [hunk 9]
```diff
@@ -232,4 +235,55 @@ impl Fixer {
...
+                }
+            }
+        }
+        node.get_text(db).to_string()
+    }
+
+    pub fn extract_variable(binary_expr: &ExprBinary, db: &dyn SyntaxGroup) -> String {
+        let lhs = binary_expr.lhs(db);
+        lhs.as_syntax_node().get_text(db).to_string()
+    }
```
Comment (0xLucqs) [inline crates/cairo-lint-core/src/fix.rs:288]: I don't think this belongs here, add it in the double comparison file

File: crates/cairo-lint-core/src/plugin.rs [hunk 10]
```diff
@@ -29,10 +30,18 @@ pub fn diagnostic_kind_from_message(message: &str) -> CairoLintKind {
...
     }
 }
 
+impl CairoLint {
+    const DESTRUCT_MATCH: &'static str =
+        "you seem to be trying to use `match` for destructuring a single pattern. Consider using `if let`";
+    const MATCH_FOR_EQUALITY: &'static str =
+        "you seem to be trying to use `match` for an equality check. Consider using `if`";
+}
+
```
Comment (0xLucqs) [inline crates/cairo-lint-core/src/plugin.rs:44]: remove that
```suggestion
```

File: crates/cairo-lint-core/src/lints/double_comparison.rs [hunk 11]
```diff
@@ -0,0 +1,38 @@
+use cairo_lang_defs::plugin::PluginDiagnostic;
+use cairo_lang_diagnostics::Severity;
+use cairo_lang_syntax::node::ast::{BinaryOperator, Expr};
+use cairo_lang_syntax::node::db::SyntaxGroup;
+use cairo_lang_syntax::node::{TypedStablePtr, TypedSyntaxNode};
+
+pub const DOUBLE_COMPARISON: &str = "redundant double comparison found. Consider simplifying to a single comparison.";
+
+pub fn check_double_comparison(db: &dyn SyntaxGroup, expr: &Expr, diagnostics: &mut Vec<PluginDiagnostic>) {
+    if let Expr::Binary(binary_op) = expr {
```
Comment (0xLucqs) [inline crates/cairo-lint-core/src/lints/double_comparison.rs:10]: you need to check that the 2 variables compared are the same for the 2 comparisons

File: crates/cairo-lint-core/src/fix.rs [hunk 12]
```diff
@@ -232,4 +235,55 @@ impl Fixer {
...
+            let rhs = binary_op.rhs(db);
+
+            if let (Expr::Binary(lhs_inner), Expr::Binary(rhs_inner)) = (&lhs, &rhs) {
+                let lhs_var = Self::extract_variable(lhs_inner, db);
+                let rhs_var = Self::extract_variable(rhs_inner, db);
+
+                if lhs_var == rhs_var {
+                    let op = binary_op.op(db);
+
+                    let simplified_op = match (op, lhs_inner.op(db), rhs_inner.op(db)) {
```
Comment (0xLucqs) [inline crates/cairo-lint-core/src/fix.rs:253]: extract this match in a function that you put in the `check_double_comparison` file and you can call it here

File: crates/cairo-lint-core/tests/test_files/double_comparison/double_comparison [hunk 13]
```diff
@@ -0,0 +1,416 @@
+//! > complex double comparison
+
+//! > cairo_code
+fn main() -> bool {
+    let x = 5;
```
Comment (0xLucqs) [inline crates/cairo-lint-core/tests/test_files/double_comparison/double_comparison:5]:
```suggestion
    let x = 5_u32;
```

File: crates/cairo-lint-core/tests/test_files/double_comparison/double_comparison [hunk 14]
```diff
@@ -0,0 +1,416 @@
...
+    }
+}
+
+//! > ==========================================================================
+
+//! > double comparison with arithmetic
+
+//! > cairo_code
+fn main() -> bool {
+    let x = 5;
```
Comment (0xLucqs) [inline crates/cairo-lint-core/tests/test_files/double_comparison/double_comparison:44]: do that for every test
```suggestion
    let x = 5_u32;
```

File: crates/cairo-lint-core/tests/test_files/double_comparison/double_comparison [hunk 15]
```diff
@@ -0,0 +1,416 @@
...
+ --> lib.cairo:5:24
+    if (x + z) == y || (x + z) < y {
+                       ^*********^
+
+//! > fixed
+fn main() -> bool {
+    let x = 5;
+    let y = 10;
+    let z = 3;
+    if (x + z) == y || (x + z) < y {
```
Comment (0xLucqs) [inline crates/cairo-lint-core/tests/test_files/double_comparison/double_comparison:31]: this should be fixed

File: crates/cairo-lint-core/src/fix.rs [hunk 16]
```diff
@@ -232,4 +235,55 @@ impl Fixer {
...
+
+                        _ => return node.get_text(db).to_string(),
+                    };
+
+                    return format!(
+                        "{}{} {} {}\n",
+                        node.get_text(db).chars().take_while(|c| c.is_whitespace()).collect::<String>(),
+                        lhs_var,
+                        simplified_op,
+                        lhs_inner.rhs(db).as_syntax_node().get_text_without_trivia(db)
```
Comment (0xLucqs) [inline crates/cairo-lint-core/src/fix.rs:277]:
```suggestion
                        rhs_var
```

**Review by 0xLucqs — COMMENTED** [review]
There's something wrong somewhere none of the tests are fixed

**Review by 0xLucqs — CHANGES_REQUESTED** [review]
None of the tests are fixed, you'll need to debug this. If you need/want some help don't hesitate

File: crates/cairo-lint-core/src/fix.rs [hunk 17]
```diff
@@ -142,6 +144,10 @@ impl Fixer {
                 self.fix_double_parens(db.upcast(), plugin_diag.stable_ptr.lookup(db.upcast()))
             }
             CairoLintKind::DestructMatch => self.fix_destruct_match(db, plugin_diag.stable_ptr.lookup(db.upcast())),
+            CairoLintKind::DoubleComparison => {
+                Self::fix_double_comparison(db.upcast(), plugin_diag.stable_ptr.lookup(db.upcast()))
+            }
+            _ => "".to_owned(),
```
Comment (0xLucqs) [inline crates/cairo-lint-core/src/fix.rs:150]:
```suggestion
```

File: crates/cairo-lint-core/src/plugin.rs [hunk 18]
```diff
@@ -81,6 +83,11 @@ impl AnalyzerPlugin for CairoLint {
                         &AstExpr::from_syntax_node(db.upcast(), node),
                         &mut diags,
                     ),
+                    SyntaxKind::ExprBinary => double_comparison::check_double_comparison(
+                        db.upcast(),
+                        &Expr::from_syntax_node(db.upcast(), descendant),
```
Comment (0xLucqs) [inline crates/cairo-lint-core/src/plugin.rs:88]: you can directly create a Binary expression with `Binary::from_syntax_node`

File: crates/cairo-lint-core/tests/test_files/double_comparison/double_comparison [hunk 19]
```diff
@@ -0,0 +1,253 @@
+//! > double comparison equal and greater than
```
Comment (0xLucqs) [inline crates/cairo-lint-core/tests/test_files/double_comparison/double_comparison:1]: add a test like that:
```rs
fn main() -> bool {
    let x = 5_u32;
    let y = 10_u32;
    let z = 1_u32;
    if x == y || z > y {
        true
    } else {
        false
    }
```

File: crates/cairo-lint-core/src/lints/double_comparison.rs [hunk 20]
```diff
@@ -0,0 +1,60 @@
...
+        (BinaryOperator::EqEq(_), BinaryOperator::LT(_))
+            | (BinaryOperator::LT(_), BinaryOperator::EqEq(_))
+            | (BinaryOperator::EqEq(_), BinaryOperator::GT(_))
+            | (BinaryOperator::GT(_), BinaryOperator::EqEq(_))
+            | (BinaryOperator::GE(_), BinaryOperator::GT(_))
+            | (BinaryOperator::LE(_), BinaryOperator::LT(_))
+            | (BinaryOperator::LE(_), BinaryOperator::GE(_))
+            | (BinaryOperator::GE(_), BinaryOperator::LE(_))
+            | (BinaryOperator::LT(_), BinaryOperator::GT(_))
+            | (BinaryOperator::GT(_), BinaryOperator::LT(_))
```
Comment (0xLucqs) [inline crates/cairo-lint-core/src/lints/double_comparison.rs:38]: you also need to check the operator in the middle because for example:
`x < y && x > y` is just invalid

File: crates/cairo-lint-core/src/lints/double_comparison.rs [hunk 21]
```diff
@@ -0,0 +1,60 @@
...
+    )
+}
+
+pub fn extract_binary_operator(expr: &Expr, db: &dyn SyntaxGroup) -> Option<BinaryOperator> {
+    if let Expr::Binary(binary_op) = expr { Some(binary_op.op(db)) } else { None }
+}
+
+pub fn extract_variable(binary_expr: &ExprBinary, db: &dyn SyntaxGroup) -> String {
+    let lhs = binary_expr.lhs(db);
+    lhs.as_syntax_node().get_text(db).to_string()
```
Comment (0xLucqs) [inline crates/cairo-lint-core/src/lints/double_comparison.rs:48]: if you just want the variable name you should use `get_text_without_trivia` and i don't think the `to_string` is necessary. Run clippy locally `cargo run clippy` it'll help you fix all the issues

**Review by 0xLucqs — CHANGES_REQUESTED** [review]

File: crates/cairo-lint-core/src/lints/double_comparison.rs [hunk 22]
```diff
@@ -0,0 +1,140 @@
...
+}
+
+pub fn operator_to_replace(lhs_op: BinaryOperator) -> &'static str {
+    match lhs_op {
+        BinaryOperator::EqEq(_) => "==",
+        BinaryOperator::GT(_) => ">",
+        BinaryOperator::LT(_) => "<",
+        BinaryOperator::GE(_) => ">=",
+        BinaryOperator::LE(_) => "<=",
+        _ => "",
```
Comment (0xLucqs) [inline crates/cairo-lint-core/src/lints/double_comparison.rs:112]: don't do that, either return an option or panic here

File: crates/cairo-lint-core/tests/test_files/double_comparison/double_comparison [hunk 23]
```diff
@@ -0,0 +1,401 @@
+//! > contradictory equal and less than
```
Comment (0xLucqs) [inline crates/cairo-lint-core/tests/test_files/double_comparison/double_comparison:1]: maybe i missed but can you also add a test where you invert x and y, for example. `x >= y && x y >= x`

File: crates/cairo-lint-core/src/lints/double_comparison.rs [hunk 24]
```diff
@@ -0,0 +1,140 @@
...
+    }
+}
+
+pub fn determine_simplified_operator(
+    lhs_op: &BinaryOperator,
+    rhs_op: &BinaryOperator,
+    middle_op: &BinaryOperator,
+) -> Option<&'static str> {
+    match (lhs_op, rhs_op, middle_op) {
```
Comment (0xLucqs) [inline crates/cairo-lint-core/src/lints/double_comparison.rs:121]: put the operator in the middle so it's easier to read

**Review by 0xLucqs — CHANGES_REQUESTED** [review]

**Review by 0xLucqs — CHANGES_REQUESTED** [review]

**Review by 0xLucqs — APPROVED** [review]
